server/sonar-server/src/test/java/org/sonar/server/platform/ws/ServerMigrateWsActionTest.java

```diff
@@ -0,0 +1,157 @@
+        when(serverUpgradeStatus.isUpgraded()).thenReturn(false);
+        when(dialect.supportsMigration()).thenReturn(true);
+        when(databaseMigration.status()).thenReturn(FAILED);
+        when(databaseMigration.startedAt()).thenReturn(SOME_DATE);
+        when(databaseMigration.failureError()).thenReturn(new UnsupportedOperationException(SOME_THROWABLE_MSG));
+
+        underTest.handle(request, response);
+
+        assertJson(response.outputAsString()).isSimilarTo(expectedResponse(false, FAILED, failedMsg(SOME_THROWABLE_MSG), SOME_DATE));
+    }
+
+    @Test
+    public void msg_is_not_operational_and_state_from_databasemigration_and_msg_has_default_msg_when_dbmigration_status_is_FAILED() throws Exception {
+        when(serverUpgradeStatus.isUpgraded()).thenReturn(false);
+        when(dialect.supportsMigration()).thenReturn(true);
+        when(databaseMigration.status()).thenReturn(FAILED);
+        when(databaseMigration.startedAt()).thenReturn(SOME_DATE);
+        when(databaseMigration.failureError()).thenReturn(null); // no failure throwable caught
+
+        underTest.handle(request, response);
+
+        assertJson(response.outputAsString()).isSimilarTo(expectedResponse(false, FAILED, failedMsg(DEFAULT_ERROR_MSG), SOME_DATE));
+    }
+
+    @Test
+    public void msg_is_operational_and_state_from_databasemigration_and_msg_has_default_msg_when_dbmigration_status_is_FAILED() throws Exception {
+        when(serverUpgradeStatus.isUpgraded()).thenReturn(false);
+        when(dialect.supportsMigration()).thenReturn(true);
+        when(databaseMigration.status()).thenReturn(SUCCEEDED);
+        when(databaseMigration.startedAt()).thenReturn(SOME_DATE);
+
+        underTest.handle(request, response);
+
+        assertJson(response.outputAsString()).isSimilarTo(expectedResponse(true, SUCCEEDED, MIG_SUCCESS_MSG, SOME_DATE));
+    }
+
+    @Test
+    public void start_migration_and_return_msg_is_not_operational_and_state_from_databasemigration_when_dbmigration_status_is_NONE() throws Exception {
+        when(serverUpgradeStatus.isUpgraded()).thenReturn(false);
+        when(dialect.supportsMigration()).thenReturn(true);
+        when(databaseMigration.status()).thenReturn(NONE).thenReturn(RUNNING);
+        when(databaseMigration.startedAt()).thenReturn(SOME_DATE);
+
+        underTest.handle(request, response);
+
+        verify(databaseMigration).startIt();
+        assertJson(response.outputAsString()).isSimilarTo(expectedResponse(false, RUNNING, RUNNING_MSG, SOME_DATE));
+    }
+
+    private static String failedMsg(@Nullable String t) {
+        return "Migration failed: " + t + ".<br/> Please check logs.";
+    }
+
+    private static String expectedResponse(boolean operational, DatabaseMigration.Status status, String msg) {
+        return "{\"operational\":" + operational + ",\"state\":\"" + status + "\",\"message\":\"" + msg + "\"}";
```

Comment:
You might want to extract those in JSON files and use WsTester.Result.assertJson() for readability (though it introduces an indirection).